src/transaction/build.rs

```diff
@@ -158,107 +172,150 @@ pub struct BuildTransactionRequestForLocalSigningFuture<T: Transport> {
+    from: Address,
+    /// The password for unlocking the account.
+    password: Protected,
+    /// The options for building the transaction request to be signed. Note that
+    /// an `Option` is used here as the future must move the value.
+    options: Option<TransactionRequestOptions>,
+    /// The inner state for the locked account future.
     #[pin]
-    sign: CompatCallFuture<T, RawTransaction>,
+    state: BuildTransactionSignedWithLockedAccountState<T>,
+}
+
+/// The inner state of the future for building locally signed transactions.
+#[pin_project]
+enum BuildTransactionSignedWithLockedAccountState<T: Transport> {
+    /// The gas price is being resolved.
+    ResolvingGasPrice(#[pin] ResolveTransactionRequestGasPriceFuture<T>),
+    /// Signing the transaction with the locked account.
+    Signing(#[pin] CompatCallFuture<T, RawTransaction>),
 }
 
 impl<T: Transport> BuildTransactionSignedWithLockedAccountFuture<T> {
     /// Create a new future for building a locally singed transaction request
     /// from a partial transaction object and account information.
     pub fn new(
-        web3: &Web3<T>,
+        web3: Web3<T>,
         from: Address,
         password: Protected,
+        gas_price: GasPrice,
         options: TransactionRequestOptions,
     ) -> Self {
-        let request = options.build_request(from);
-
-        let password = unsafe { str::from_utf8_unchecked(password.as_ref()) };
-        let sign = web3.personal().sign_transaction(request, password).compat();
+        let state = BuildTransactionSignedWithLockedAccountState::ResolvingGasPrice(
+            gas_price.resolve_for_transaction_request(&web3),
+        );
 
-        BuildTransactionSignedWithLockedAccountFuture { sign }
+        BuildTransactionSignedWithLockedAccountFuture {
+            web3,
+            from,
+            password,
+            options: Some(options),
+            state,
+        }
     }
 }
 
 impl<T: Transport> Future for BuildTransactionSignedWithLockedAccountFuture<T> {
     type Output = Result<Bytes, ExecutionError>;
 
+    #[project]
     fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-        self.project().sign.poll(cx).map(|raw| Ok(raw?.raw))
+        let mut this = self.project();
+
+        loop {
+            #[project]
+            let next_state = match this.state.as_mut().project() {
+                BuildTransactionSignedWithLockedAccountState::ResolvingGasPrice(gas_price) => {
+                    let gas_price = match ready!(gas_price.poll(cx)).transpose() {
+                        Ok(gas_price) => gas_price,
+                        Err(err) => return Poll::Ready(Err(err.into())),
+                    };
+
+                    let options = this.options.take().expect("future called more than once");
+                    let request = options.build_request(*this.from, gas_price);
+                    let password = unsafe { str::from_utf8_unchecked(this.password.as_ref()) };
```

Comment:
🙁 is there another reason to use unsafe here besides trying to avoid the validity check of `str::from_utf8`? I'd prefer having that micro performance hit in order to avoid setting precedence (I must have overseen it before) of unsafe code.

Reply:
The unsafe was there before but I don't remember my reasoning. It definitely does not need to be there at all.

I think it might even have been a "I'll get back to this later and pipe through the error" but forgot.

Reply:
I'll do this as a separate PR.